File: app/src/main/java/com/davismiyashiro/expenses/view/opentab/ParticipantRecyclerViewAdapter.java
package com.davismiyashiro.expenses.view.opentab;

import android.content.Context;
import android.content.DialogInterface;
import android.support.v7.app.AlertDialog;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import com.davismiyashiro.expenses.R;
import com.davismiyashiro.expenses.datatypes.Participant;
import com.davismiyashiro.expenses.view.opentab.ParticipantFragment.OnParticipantListFragmentInteractionListener;

import java.util.List;

/**
 * {@link RecyclerView.Adapter} that can display an item and makes a call to the
 * specified {@link OnParticipantListFragmentInteractionListener}.
 */
public class ParticipantRecyclerViewAdapter extends RecyclerView.Adapter<ParticipantRecyclerViewAdapter.ViewHolder> {

    private List<Participant> mParticipants;
    private final OnParticipantListFragmentInteractionListener mListener;
    private Context mContext;

    public ParticipantRecyclerViewAdapter(List<Participant> participants, OnParticipantListFragmentInteractionListener listener, Context context) {
        mParticipants = participants;
        mListener = listener;
        mContext = context;
    }

    @Override
    public ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(parent.getContext())
                .inflate(R.layout.fragment_participant, parent, false);
        return new ViewHolder(view);
    }

    @Override
    public void onBindViewHolder(final ViewHolder holder, int position) {
        holder.mItem = mParticipants.get(position);
        holder.mIdView.setText(mParticipants.get(position).getName());
        holder.mContentView.setText(mParticipants.get(position).getEmail());

//        holder.mView.setOnClickListener(new View.OnClickListener() {
//            @Override
//            public void onClick(View v) {
//                if (null != mListener) {
//                    // Notify the active callbacks interface (the activity, if the
//                    // fragment is attached to one) that an item has been selected.
//                    mListener.onParticipantListFragmentInteraction(holder.mItem);
//                }
//            }
//        });
        //return true to indicate that you have handled the event and it should stop here; return false if you have not handled it and/or the event should continue to any other on-click listeners.
//        holder.mView.setOnLongClickListener(new View.OnLongClickListener() {
//            @Override
//            public boolean onLongClick(View v) {
//
//                return true;
//            }
//        });
    }

    public void replaceData (List<Participant> participants) {
        setParticipants(participants);
        notifyDataSetChanged();
    }

    private void setParticipants (List<Participant> participants) {
        mParticipants = participants;
    }

    @Override
    public int getItemCount() {
        return mParticipants.size();
    }

    public Participant getItem(int position) {
        return mParticipants.get(position);
    }

    public void removeItemAtPosition(int position) {
        mParticipants.remove(position);
        notifyItemRemoved(position);
    }

    class ViewHolder extends RecyclerView.ViewHolder {
        final View mView;
        final TextView mIdView;
        final TextView mContentView;
        Participant mItem;

        ViewHolder(View view) {
            super(view);
            mView = view;
            mIdView = (TextView) view.findViewById(R.id.list_item_title);
            mContentView = (TextView) view.findViewById(R.id.list_item_details);

            mView.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    if (null != mListener) {
                        // Notify the active callbacks interface (the activity, if the
                        // fragment is attached to one) that an item has been selected.
                        mListener.onParticipantListFragmentInteraction(mItem);
                    }
                }
            });

            mView.setOnLongClickListener(new View.OnLongClickListener() {
                @Override
                public boolean onLongClick(View v) {
                    AlertDialog.Builder alertDialogBuilder = new AlertDialog.Builder(mContext);
                    alertDialogBuilder.setTitle("Delete Participant");
                    alertDialogBuilder.setMessage("Are you sure you want to delete this Participant ?");

                    alertDialogBuilder.setPositiveButton("YES", new DialogInterface.OnClickListener() {
                        @Override
                        public void onClick(DialogInterface arg0, int arg1) {
                            int position = getAdapterPosition();

                            mListener.onParticipantListFragmentLongClick(getItem(position));
                            removeItemAtPosition(position);
                            arg0.dismiss();
                        }
                    });

                    alertDialogBuilder.setNegativeButton("NO",new DialogInterface.OnClickListener() {
                        @Override
                        public void onClick(DialogInterface dialog, int which) {
                            dialog.cancel();
                        }
                    });

                    AlertDialog alertDialog = alertDialogBuilder.create();
                    alertDialog.show();

                    return true;
                }
            });
        }

        @Override
        public String toString() {
            return super.toString() + " '" + mIdView.getText() + "'";
        }
    }
}

File: app/src/main/java/com/davismiyashiro/expenses/view/opentab/ParticipantRecyclerViewAdapter.kt
/*
 * MIT License
 *
 * Copyright (c) 2019 Davis Miyashiro
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
package com.davismiyashiro.expenses.view.opentab

import android.content.Context
import android.support.v7.app.AlertDialog
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView

import com.davismiyashiro.expenses.R
import com.davismiyashiro.expenses.datatypes.Participant

/**
 * [RecyclerView.Adapter] that can display an item and makes a call to the
 * specified [OnParticipantListFragmentInteractionListener].
 */
class ParticipantRecyclerViewAdapter(
    private var mParticipants: MutableList<Participant>,
    private val mListener: OnParticipantListFragmentInteractionListener?,
    private val mContext: Context
) : RecyclerView.Adapter<ParticipantRecyclerViewAdapter.ViewHolder>() {

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val view = LayoutInflater.from(parent.context)
                .inflate(R.layout.fragment_participant, parent, false)
        return ViewHolder(view)
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        holder.mItem = mParticipants[position]
        holder.mIdView.text = mParticipants[position].name
        holder.mContentView.text = mParticipants[position].email
    }

    fun replaceData(participants: MutableList<Participant>) {
        setParticipants(participants)
        notifyDataSetChanged()
    }

    private fun setParticipants(participants: MutableList<Participant>) {
        mParticipants = participants
    }

    override fun getItemCount(): Int {
        return mParticipants.size
    }

    fun getItem(position: Int): Participant {
        return mParticipants[position]
    }

    fun removeItemAtPosition(position: Int) {
        mParticipants.removeAt(position)
        notifyItemRemoved(position)
    }

    inner class ViewHolder(mView: View) : RecyclerView.ViewHolder(mView) {
        val mIdView: TextView
        val mContentView: TextView
        lateinit var mItem: Participant

        init {
            mIdView = mView.findViewById<View>(R.id.list_item_title) as TextView
            mContentView = mView.findViewById<View>(R.id.list_item_details) as TextView

            mView.setOnClickListener { v ->
                mListener?.onParticipantListFragmentInteraction(mItem)
            }

            mView.setOnLongClickListener { v ->
                val alertDialogBuilder = AlertDialog.Builder(mContext)
                alertDialogBuilder.setTitle("Delete Participant")
                alertDialogBuilder.setMessage("Are you sure you want to delete this Participant ?")

                alertDialogBuilder.setPositiveButton("YES") { arg0, arg1 ->
                    val position = adapterPosition

                    mListener?.onParticipantListFragmentLongClick(getItem(position))
                    removeItemAtPosition(position)
                    arg0.dismiss()
                }

                alertDialogBuilder.setNegativeButton("NO") { dialog, which -> dialog.cancel() }

                val alertDialog = alertDialogBuilder.create()
                alertDialog.show()

                true
            }
        }

        override fun toString(): String {
            return super.toString() + " '" + mIdView.text + "'"
        }
    }

    /**
     * Best practices suggests that Fragments should have interfaces to be implemented by the Activities
     * that contain this Fragment. Although as functionality grows one presenter gets overwhelmed, so
     * in order to provide a cleaner separation of concerns, a presenter for each Fragment was used.
     *
     * More info on SharedPresenters/ViewModels:
     * See the Android Training lesson [Communicating with Other Fragments](http://developer.android.com/training/basics/fragments/communicating.html) for more information.
     */
    interface OnParticipantListFragmentInteractionListener {
        fun onParticipantListFragmentInteraction(item: Participant)

        fun onParticipantListFragmentLongClick(item: Participant)
    }
}
